ocaml/vhd-tool/src/nbd_input.ml

```diff
@@ -103,3 +103,62 @@ let raw ?(extent_reader = "/opt/xensource/libexec/get_nbd_extents.py") raw
   block [] 0L >>= fun elements ->
   let size = Vhd_format.F.{total= size; metadata= 0L; empty= 0L; copy= 0L} in
   Lwt.return F.{elements; size}
+
+let vhd ?(extent_reader = "/opt/xensource/libexec/get_nbd_extents.py") raw
+    server export_name size =
+  (* All units are bytes *)
+  let open Int64 in
+  (* Get extents directly from the NBD server through a helper program.
+     Each extent has a length and flags that indicate whether or not there is data. *)
+  let rec read_extents extents offset =
+    if offset > size then
+      Lwt.fail_with
+        (Printf.sprintf "Nbd_input.vhd finished with offset=%Ld <> size=%Ld"
+           offset size
+        )
+    else if offset = size then
+      Lwt.return (List.rev extents)
+    else
+      let length = min (sub size offset) max_query_length in
+      get_extents_json ~extent_reader ~server ~export_name ~offset ~length
+      >>= fun extents_json ->
+      let new_extents = extent_list_of_rpc (Jsonrpc.of_string extents_json) in
+      read_extents (List.rev_append new_extents extents) (add offset length)
+  in
+  read_extents [] 0L >>= fun extents ->
+  let find_data_blocks ~blocks:_ ~block_size =
+    (* Given a list of extents and a VHD block size, return a list of block indices
+       for all blocks that contain data. *)
+    let add_new i l =
+      (* Add i to list l, unless the list's head is already equal to i *)
+      match l with [] -> [i] | j :: _ when i = j -> l | _ -> i :: l
+    in
+    let rec loop offset acc = function
+      | e :: es ->
```

Comment:
We'd still have to do the recursive call in the empty-extent case, with updated offset and sector sanity check. But it may overall be clearer – I'll have a go.